accepted/isle-extended-patterns.md

```diff
@@ -0,0 +1,425 @@
+  left-hand side, as extractors. In return, the ISLE compiler (and
+  hence really the user) does not have to worry about arbitrary-depth
+  backtracking: once the left-hand side matches, we are
+  committed.[^2][^3] This is a desirable complexity-limiting
+  constraint.
+  
+[^2]: It is interesting to note that the use of side-effects to emit
+      instructions for the output is actually something that underwent
+      several changes during the design of the DSL, as well. The very
+      early ideas for ISLE were essentially a Prolog-like system where
+      side effects were explicitly used. We later determined that a
+      "value-oriented" system, where the term rewriting at least
+      conceptually rewrites values to equivalent values, was easier to
+      manage; but in order not to refactor the whole VCode design at
+      the same time, the returned value does not contain a tree or DAG
+      of instructions, but rather just references to them (via
+      `Reg`s), so side-effects eventually made their appearance anyway
+      as we emit them via the side-effecting `emit` constructor. This
+      is a property of the Cranelift glue and not necessarily
+      fundamental to the DSL itself, but seems to be a reasonable
+      design point that we should optimize for regardless.
+
+[^3]: RHS is actually fallible because the ISLE execution as a whole
+      is allowed to be "partial", so that we can gradually migrate to
+      the language and fall back to legacy code. This is why
+      constructors return `Option<T>`; the `Option` is for a different
+      reason than the fallible extractors' `Option` that carries
+      semantic meaning ("did I match or not").
+      
+- The lack of backtracking also means that we do not need any sort of
+  "undo" of constructors, and constructors are free to have incidental
+  side-effects (such as emitting instructions so that we can return
+  their result registers, or allocating temporaries) even as the
+  top-level form of the rules is one of rewriting equivalent values.
+  
+The above properties strongly argue that we should retain a LHS-to-RHS
+rewriting-rule structure, rather than (for example) lowering ISLE to a
+more generic Prolog-like form where extractors and constructors are
+unified.
+
+Given that constraint, our key proposal is laid out in two parts:
+extended LHS patterns with `given` clauses, and a special notion of
+pure expressions.
+
+### Left-hand-side `given` clauses
+
+First, we propose to extend the `(rule ...)` form to take a list of
+LHS clauses. The current form, `(rule LHS RHS)`, applies the pattern
+`LHS` to the value to be rewritten. We extend this with:
+  
+```lisp
+
+(rule LHS_PATTERN
+  (given PAT2 EXPR2)
+  (given PAT3 EXPR3)
+  ...
+  RHS)
+```
+  
+This means: match the input value with `LHS_PATTERN`. Then evaluate
+`EXPR2` and match its result with `PAT2`. Likewise down the list,
+until we have matched all `given` clauses.
```

Comment:
Can we expand on this a bit? I.e., if any `given` clause is unable to match, the rule does not match? Also, including the case of a boolean predicate more explicitly here?

Reply:
Expanded this with a more explicit discussion of how the evaluation proceeds.